sweepbatcher/sweep_batch.go

```diff
@@ -0,0 +1,1390 @@
+	b.log.Infof("confirmed")
+	b.state = Confirmed
+
+	// Give the signal back to the batcher that this batch is now completed.
+	// We do not care about the outcome thus we don't block for it.
+	go func() {
+		b.completeChan <- b.id
+	}()
+
+	return b.store.ConfirmBatch(ctx, b.id)
+}
+
+// persist updates the batch in the database.
+func (b *batch) persist(ctx context.Context) error {
+	bch := &dbBatch{}
+
+	bch.ID = b.id
+	bch.State = stateEnumToString(b.state)
+
+	if b.batchTxid != nil {
+		bch.BatchTxid = *b.batchTxid
+	}
+
+	bch.BatchPkScript = b.batchPkScript
+	bch.LastRbfHeight = b.rbfCache.LastHeight
+	bch.LastRbfSatPerKw = int32(b.rbfCache.FeeRate)
+	bch.MaxTimeoutDistance = b.Cfg.MaxTimeoutDistance
+
+	return b.store.UpdateSweepBatch(ctx, bch)
+}
+
+// addSweep writes a sweep to the batch's request channel.
+func (b *batch) addSweep(ctx context.Context, sweep *sweep) error {
+	select {
+	case b.requestChan <- sweep:
+		return nil
+
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// getBatchDestAddr returns the batch's destination address. If the batch
+// has already generated an address then the same one will be returned.
+func (b *batch) getBatchDestAddr(ctx context.Context) (btcutil.Address, error) {
+	var address btcutil.Address
+
+	// If a batch address is set, use that. Otherwise, generate a
+	// new address.
+	if b.batchAddress != nil {
+		address = b.batchAddress
+	} else {
+		var err error
+
+		// Generate a wallet address for the batch transaction's output.
+		address, err = b.wallet.NextAddr(
+			ctx, "", walletrpc.AddressType_TAPROOT_PUBKEY, false,
+		)
+		if err != nil {
+			return address, err
+		}
+
+		// Save that new address in order to re-use in future
+		// versions of the batch tx.
+		b.batchAddress = address
+	}
+
+	return address, nil
+}
+
+func (b *batch) insertAndAcquireID(ctx context.Context) (int32, error) {
+	bch := &dbBatch{}
+	bch.State = stateEnumToString(b.state)
+	bch.MaxTimeoutDistance = b.Cfg.MaxTimeoutDistance
+
+	id, err := b.store.InsertSweepBatch(ctx, bch)
+	if err != nil {
+		return 0, err
+	}
+
+	b.id = id
+	b.log = batchPrefixLogger(fmt.Sprintf("%d", b.id))
+
+	return id, nil
+}
+
+// notifySweepSpend writes the spendTx to the sweep's notifier channel.
+func notifySweepSpend(ctx context.Context, s sweep, spendTx *wire.MsgTx) {
+	select {
+	// Try to write the update to the notification channel.
+	case s.notifier.SpendChan <- spendTx:
+
+	// If a quit signal was provided by the swap, continue.
+	case <-s.notifier.QuitChan:
+
+	// If the context was canceled, return.
+	case <-ctx.Done():
+	}
+}
+
+// purgeSweep writes the sweep request back to the batcher in order for it to be
+// handled again.
+func purgeSweep(ctx context.Context, s SweepRequest,
+	returnChan chan<- SweepRequest) {
+
+	select {
+	case returnChan <- s:
+	case <-ctx.Done():
+	}
+}
+
+func (b *batch) writeToErrChan(ctx context.Context, err error) {
+	select {
+	case b.errChan <- err:
+	case <-ctx.Done():
```

Comment:
This is dangerous, as go does not guarantee that `b.errChan <- err` will be attempted first, and it could directly jump to `default`, leading to a no-op. Check this [`link`](https://medium.com/a-journey-with-go/go-ordering-in-select-statements-fd0ff80fd8d6)

Also if the `errChan` is already full, and we have another error on top of that for delivery I don't think skipping it is going to help as that's just loss of information. It's better to wait for the first error to get handled and then (if that doesn't cause a daemon exit) wait for the new one to be handled as well.

Reply:
> This is dangerous, as go does not guarantee that b.errChan <- err will be attempted first, and it could directly jump to default, leading to a no-op. Check this [link](https://medium.com/a-journey-with-go/go-ordering-in-select-statements-fd0ff80fd8d6)

Not true.
See spec: https://go.dev/ref/spec#Select_statements
```
1) For all the cases in the statement, the channel operands of receive operations and the channel and right-hand-side expressions of send statements are evaluated exactly once, in source order, upon entering the "select" statement. The result is a set of channels to receive from or send to, and the corresponding values to send. Any side effects in that evaluation will occur irrespective of which (if any) communication operation is selected to proceed. Expressions on the left-hand side of a RecvStmt with a short variable declaration or assignment are not yet evaluated.
1) If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection. Otherwise, if there is a default case, that case is chosen. If there is no default case, the "select" statement blocks until at least one of the communications can proceed.
3) Unless the selected case is the default case, the respective communication operation is executed.
```

> Also if the errChan is already full, and we have another error on top of that for delivery I don't think skipping it is going to help as that's just loss of information

The reaction to pushing an error to the error channel is that we quit the event loop, so no need to process all errors, only the first one.

Reply:
> 1) If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection.

Ok I thought the pseudo-random selection included `default`, my case would stand if we wanted order over multiple channel reads. :+1: